app/api/common.v2/database/MongoDataSource.ts

```diff
@@ -1,28 +1,132 @@
-import { Db, Document } from 'mongodb';
+import { Collection, Db, Document } from 'mongodb';
 import { BulkWriteStream } from './BulkWriteStream';
 import { MongoTransactionManager } from './MongoTransactionManager';
 
+type MaxAmountOfParameters<F extends (...args: any) => any> = Required<Parameters<F>>['length'];
+
+/**
+ * Map of collection members to position of its options argument.
+ * Position is based on 1.
+ */
+type MethodsOptionsArgPosition = {
+  [method in keyof Collection]: Collection[method] extends (...args: any) => any
+    ? MaxAmountOfParameters<Collection[method]>
+    : null;
+};
+
 export abstract class MongoDataSource<CollectionSchema extends Document = any> {
   protected db: Db;
 
   protected abstract collectionName: string;
 
   private transactionManager: MongoTransactionManager;
 
+  private collectionProxy?: Collection<CollectionSchema>;
+
   constructor(db: Db, transactionManager: MongoTransactionManager) {
     this.db = db;
     this.transactionManager = transactionManager;
   }
 
+  static scopedMethods: MethodsOptionsArgPosition = {
+    insertOne: 2,
+    insertMany: 2,
+    bulkWrite: 2,
+    updateOne: 3,
+    replaceOne: 3,
+    updateMany: 3,
+    deleteOne: 2,
+    deleteMany: 2,
+    rename: 2,
+    drop: 1,
+    findOne: 2,
+    find: 2,
+    options: 1,
+    isCapped: 1,
+    createIndex: 2,
+    createIndexes: 2,
+    dropIndex: 2,
+    dropIndexes: 1,
+    listIndexes: 1,
+    indexExists: 2,
+    indexInformation: 1,
+    estimatedDocumentCount: 1,
+    countDocuments: 2,
+    distinct: 3,
+    indexes: 1,
+    stats: 1,
+    findOneAndDelete: 2,
+    findOneAndReplace: 3,
+    findOneAndUpdate: 3,
+    aggregate: 2,
+    watch: 2,
+    initializeUnorderedBulkOp: 1,
+    initializeOrderedBulkOp: 1,
+    count: 2,
+    dbName: null,
+    collectionName: null,
+    namespace: null,
+    readConcern: null,
+    readPreference: null,
+    bsonOptions: null,
+    writeConcern: null,
+    hint: null,
+  } as const;
+
+  private appendSessionToOptions(args: any[], method: keyof MethodsOptionsArgPosition) {
+    const optionsPosition = MongoDataSource.scopedMethods[method]! - 1;
+    const minArgumentsLength = optionsPosition;
+    const missingArguments = minArgumentsLength - args.length;
+
+    const paddedArgs =
+      missingArguments > 0 ? args.concat(Array(missingArguments).fill(undefined)) : args;
+
+    paddedArgs[optionsPosition] = {
+      ...paddedArgs[optionsPosition],
+      session: this.transactionManager.getSession(),
+    };
+
+    return paddedArgs;
+  }
+
+  private scopeCollectionToSession(collection: Collection<CollectionSchema>) {
+    if (this.collectionProxy) {
+      return this.collectionProxy;
+    }
+
+    const self = this;
```

Comment:
Yes... aliasing `this` is a bit more performant than binding, but that's not the reason whatsoever. Within the proxy, `this` is always set to the receiver (the object in which the property being proxied was called), so I didn't want to change it since I don't know what effects it could have. Probably none, but erred on the side of safety :)